### PR TITLE
Initial test set-up for datastore-entities

### PR DESCRIPTION
**Pull request by Billcountry**
- Added circle CI config
- Added Codecov config
- Added tests for reading operations including querying and getting
- Added rules for merging to master

### CODE REVIEW

**Comment by cymox1** [comment]
# [Codecov](https://codecov.io/gh/echomobi/echo-datastore-entities/pull/1?src=pr&el=h1) Report
> :exclamation: No coverage uploaded for pull request base (`master@b5708c5`). [Click here to learn what that means](https://docs.codecov.io/docs/error-reference#section-missing-base-commit).
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/echomobi/echo-datastore-entities/pull/1?src=pr&el=tree)

```diff
@@            Coverage Diff            @@
##             master       #1   +/-   ##
=========================================
  Coverage          ?   42.72%           
=========================================
  Files             ?        4           
  Lines             ?      110           
  Branches          ?        0           
=========================================
  Hits              ?       47           
  Misses            ?       63           
  Partials          ?        0           
```



------

[Continue to review full report at Codecov](https://codecov.io/gh/echomobi/echo-datastore-entities/pull/1?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/echomobi/echo-datastore-entities/pull/1?src=pr&el=footer). Last update [b5708c5...d358c8c](https://codecov.io/gh/echomobi/echo-datastore-entities/pull/1?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: tests/test_query.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        entities = TestEntity.query().lte("prop2", 20).fetch()
+        self.assertEqual(len(entities), 21)
+        for entity in entities:
+            self.assertTrue(entity.prop2 <= 20)
+
+    def test_compound_query(self):
+        entities = TestEntity.query().gte("prop2", 20).lt("prop2", 30).fetch()
+        self.assertEqual(len(entities), 10)
+        for entity in entities:
+            self.assertTrue(20 <= entity.prop2 < 30)
```
Comment (georgemarrows) [inline tests/test_query.py]: I didn't know you could do `20 <= entity.prop2 < 30` in Python! Cool.

File: tests/test_entity.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        entity = TestEntity(id=10)
+        project = Entity.__get_client__().project
+        expected_key = Key('TestEntity', 10, project=project)
+        self.assertEqual(str(entity.key()), expected_key.to_legacy_urlsafe().decode("utf-8"))
+        self.assertEqual(entity.key().id, 10)
+        self.assertEqual(expected_key.id, entity.key().id)
+        # TODO: Test for key generation after put
+
+    def test_get(self):
```
[the remark this answers is not in the document]
Comment (georgemarrows) [reply tests/test_entity.py]: Oh hold on, looks like wrapper's `put` doesn't work yet.

[the remark this answers is not in the document]
Comment (Billcountry) [reply tests/test_entity.py]: It's a todo coz the library has no write capabilities at the moment. Just read

File: tests/test_entity.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    prop2 = db.IntegerProperty()
+
+
+class TestEntityTestCase(unittest.TestCase):
+    def test_client(self):
+        self.assertIsInstance(Entity.__get_client__(), Client)
+
+    def test_value_setting(self):
+        entity = TestEntity()
+        entity.prop1 = "Text Value"
```
[the remark this answers is not in the document]
Comment (Billcountry) [reply tests/test_entity.py]: I'll add one. Should throw invalid value error.

File: tests/test_entity.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        # You should not get a key for an unsaved entity
+        self.assertRaises(errors.NotSavedException, entity.key)
+
+        entity = TestEntity(id=10)
+        project = Entity.__get_client__().project
+        expected_key = Key('TestEntity', 10, project=project)
+        self.assertEqual(str(entity.key()), expected_key.to_legacy_urlsafe().decode("utf-8"))
+        self.assertEqual(entity.key().id, 10)
+        self.assertEqual(expected_key.id, entity.key().id)
+        # TODO: Test for key generation after put
```
[the remark this answers is not in the document]
Comment (Billcountry) [reply tests/test_entity.py]: Once we have implemented write into the library

File: tests/test_entity.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        entity = TestEntity()
+        entity.prop1 = "Text Value"
+        entity.prop2 = 1
+        self.assertEqual(entity.__datastore_entity__.get("prop1"), "Text Value")
+        self.assertEqual(entity.__datastore_entity__.get("prop2"), 1)
+
+    def test_key_generation(self):
+        entity = TestEntity()
+        # You should not get a key for an unsaved entity
+        self.assertRaises(errors.NotSavedException, entity.key)
```
[the remark this answers is not in the document]
Comment (Billcountry) [reply tests/test_entity.py]: Nope, `assertRaises` automatically calls the method. Variables are provided as `*args, **kwargs`

File: tests/test_query.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        for entity in TestEntity.query(order_by=["-prop2"]):
+            i = 50 - count - 1
+            expected_key = self.client.key('TestEntity', i + 1).to_legacy_urlsafe().decode("utf-8")
+            self.assertEqual(str(entity.key()), expected_key)
+            self.assertEqual("Odd" if i % 2 else "Even", entity.prop1)
+            self.assertEqual(i, entity.prop2)
+            count += 1
+
+    def test_order_and_limit(self):
+        limit = randint(2, 20)
```
[the remark this answers is not in the document]
Comment (Billcountry) [reply tests/test_query.py]: Setting a fixed limit for now

File: tests/test_entity.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Billcountry) [reply tests/test_entity.py]: Done

**Comment by Billcountry** [comment]
@georgemarrows I've removed stub methods for put.
Added more docs for the methods. I'm not sure we have a way to test `eventual` at the moment.
Key_only is now tested

File: tests/test_entity.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (georgemarrows) [reply tests/test_entity.py]: Ah, wasn't thinking how `assertRaises` had to work.

File: tests/test_entity.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (georgemarrows) [reply tests/test_entity.py]: Agree. I saw later that `put` isn't yet there.

File: echo/datastore/entity.py [hunk 7]
```diff
@@ -12,45 +25,103 @@ class Entity(object):
...
         if "id" in data:
             self.__id = data.get("id")
-        self.__datastore_entity__ = DatastoreEntity(key=self.key())
+        self.__datastore_entity__ = DatastoreEntity(key=self.key(partial=True))
         for key, value in data.items():
             setattr(self, key, value)
 
-    def key(self):
+    def key(self, partial=False):
+        """Generates a key for this Entity
```
Comment (georgemarrows) [inline echo/datastore/entity.py]: Thanks for adding these comments. Are they taken from the actual Google old datastore API implementation?

Comment (Billcountry) [reply echo/datastore/entity.py]: Not this one, but some that translate directly yes. e.g. in those in Query

**Pull request merged**